Comptime branch limit reached

I'm really not sure how helpful I can really be here, but there seems to be an issue during comptime compilation (when cross compiling to 32 bit arm at the very least) where it hits the eval branch quota:
```
$ zig build
anyzig: zig version '0.15.2' pulled from '/home/wgrav/Code/firepad/build.zig.zon'
anyzig: appdata '/home/wgrav/.local/share/anyzig'
anyzig: zig '0.15.2' already exists at '/home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ'
install
└─ install client
   └─ compile exe client Debug arm-linux-gnueabihf
      └─ run exe gen_event (Event.zig)
         └─ compile exe gen_event Debug native 1 errors
/home/wgrav/.cache/zig/p/evdev-0.1.0-1sSqaKObAABeWFqMzJW0B3bkooUDLXNKpJmaQPOFIsbK/build/gen_event.zig:155:37: error: evaluation exceeded 220000 backwards branches
                comptime if (!filter(cnst, typ ++ "_"))
                              ~~~~~~^~~~~~~~~~~~~~~~~~
/home/wgrav/.cache/zig/p/evdev-0.1.0-1sSqaKObAABeWFqMzJW0B3bkooUDLXNKpJmaQPOFIsbK/build/gen_event.zig:155:37: note: use @setEvalBranchQuota() to raise the branch limit from 220000
referenced by:
    callMain [inlined]: /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/lib/std/start.zig:627:37
    callMainWithArgs [inlined]: /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/lib/std/start.zig:587:20
    main: /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/lib/std/start.zig:602:28
    1 reference(s) hidden; use '-freference-trace=4' to see all references
error: the following command failed with 1 compilation errors:
/home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/zig build-exe .zig-cache/o/2105d1fb4f6f0bc9b2b89d1437366143/libevdev.so.2.3.0 -rpath .zig-cache/o/2105d1fb4f6f0bc9b2b89d1437366143 -I .zig-cache/o/302a70f6ee7f28d9010b41eb976eb6e4 -Mroot=/home/wgrav/.cache/zig/p/evdev-0.1.0-1sSqaKObAABeWFqMzJW0B3bkooUDLXNKpJmaQPOFIsbK/build/gen_event.zig -lc --cache-dir .zig-cache --global-cache-dir /home/wgrav/.cache/zig --name gen_event --zig-lib-dir /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/lib/ --listen=-

Build Summary: 5/12 steps succeeded; 1 failed
install transitive failure
├─ install server transitive failure
│  └─ compile exe server Debug native transitive failure
│     └─ run exe gen_event (Event.zig) transitive failure
│        └─ compile exe gen_event Debug native 1 errors
└─ install client transitive failure
   └─ compile exe client Debug arm-linux-gnueabihf transitive failure
      └─ run exe gen_event (Event.zig) (+1 more reused dependencies)

error: the following build command failed with exit code 1:
.zig-cache/o/003605bf534e716fd4427de2666c99e1/build /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/zig /home/wgrav/.cache/zig/p/N-V-__8AANeVhBSiRwesZrZKZ1hTkJd-goYzJWpN6kwXi-FJ/lib /home/wgrav/Code/firepad .zig-cache /home/wgrav/.cache/zig --seed 0xe5bc9dfc -Z9baa416b18daa57e
```

And for context, my build.zig: 
```zig
// Copyright (C) 2026 wgrav
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

const std = @import("std");

pub fn build(b: *std.Build) void {
    const clap = b.dependency("clap", .{});
    const evdev = b.dependency("evdev", .{});

    const target = b.standardTargetOptions(.{});
    const optimize = b.standardOptimizeOption(.{});

    const server = b.addExecutable(.{
        .name = "server",
        .use_llvm = true, // should fix comptime issues
        .root_module = b.createModule(.{
            .root_source_file = b.path("src/server.zig"),
            .target = target,
            .optimize = optimize,
            .imports = &.{
                .{ .name = "clap", .module = clap.module("clap") },
                .{ .name = "evdev", .module = evdev.module("evdev") },
            },
        }),
    });

    b.installArtifact(server);

    const kindle = b.resolveTargetQuery(.{
        .cpu_arch = .arm,
        .os_tag = .linux,
        .abi = .gnueabihf,
        .cpu_features_add = std.Target.arm.cpu.cortex_a53.features
    });

    const client = b.addExecutable(.{
        .name = "client",
        .use_llvm = true, // ditto
        .root_module = b.createModule(.{
            .root_source_file = b.path("src/client.zig"),
            .target = kindle,
            .optimize = optimize,
            .imports = &.{
                .{ .name = "evdev", .module = evdev.module("evdev") }
            },
        }),
    });

    b.installArtifact(client);

    const server_tests = b.addTest(.{
        .root_module = server.root_module,
    });

    const run_server_tests = b.addRunArtifact(server_tests);

    const test_step = b.step("test", "Run tests");
    test_step.dependOn(&run_server_tests.step);
}
```

This is on 0.15.2, and the files themselves (`server.zig` and `client.zig`) contain only empty `main()` functions. Do let me know if there's any extra context you're missing.

Thanks for making this, though!

--- a/build/gen_event.zig
+++ b/build/gen_event.zig
@@ -17,16 +17,39 @@ pub fn lowercase(arg: []const u8) []const u8 {
         @panic("Failed to lowercase string for unknown reason");
 }
 
-pub fn main() !void {
-    @setEvalBranchQuota(220000);
-    const allocator = std.heap.page_allocator;
+fn isMetaConstant(name: []const u8) bool {
+    for ([_][]const u8{ "_VERSION", "_CNT", "_MAX" }) |s| {
+        if (std.mem.endsWith(u8, name, s)) return true;
+    }
+    return false;
+}
 
-    const consts = comptime b: {
-        var consts: []const []const u8 = &.{};
-        for (@typeInfo(c).@"struct".decls) |decl|
-            consts = consts ++ &[_][]const u8{decl.name};
-        break :b consts;
-    };
+const Constant = struct {
+    name: []const u8,
+    full_name: []const u8,
+    value: c_int,
+
+    fn collect(comptime prefix: []const u8) []const Constant {
+        return comptime b: {
+            var slice: []const Constant = &.{};
+            for (@typeInfo(c).@"struct".decls) |decl| {
+                if (!std.mem.startsWith(u8, decl.name, prefix)) continue;
+                if (isMetaConstant(decl.name)) continue;
+                const self: Constant = .{
+                    .name = decl.name[prefix.len..],
+                    .full_name = decl.name,
+                    .value = @field(c, decl.name),
+                };
+                slice = slice ++ &[_]Constant{self};
+            }
+            break :b slice;
+        };
+    }
+};
+
+pub fn main() !void {
+    @setEvalBranchQuota(30000);
+    const allocator = std.heap.page_allocator;
 
     var s: std.ArrayList(u8) = .empty;
     defer s.deinit(allocator);
@@ -40,24 +63,17 @@ pub fn main() !void {
     ) catch {};
 
     // SYN, KEY, REL, ...
-    const types = comptime b: {
-        var types: []const []const u8 = &.{};
-        for (consts) |cnst| {
-            if (filter(cnst, "EV_"))
-                types = types ++ &[_][]const u8{cnst["EV_".len..]};
-        }
-        break :b types;
-    };
+    const event_types = comptime Constant.collect("EV_");
 
     {
         _ = w.write(
             \\pub const Type = enum(c_ushort) {
             \\
         ) catch {};
-        inline for (types) |typ| w.print(
+        for (event_types) |event_type| w.print(
             \\    {s} = {},
             \\
-        , .{ lowercase(typ), @field(c, "EV_" ++ typ) }) catch {};
+        , .{ lowercase(event_type.name), event_type.value }) catch {};
         _ = w.write(
             \\    pub inline fn new(integer: c_ushort) Type {
             \\        return @enumFromInt(integer);
@@ -72,10 +88,10 @@ pub fn main() !void {
             \\        return switch (self) {
             \\
         ) catch {};
-        inline for (types) |typ| w.print(
-            \\          .{s} => "EV_{s}",
+        for (event_types) |event_type| w.print(
+            \\          .{s} => "{s}",
             \\
-        , .{ lowercase(typ), typ }) catch {};
+        , .{ lowercase(event_type.name), event_type.full_name }) catch {};
         _ = w.write(
             \\        };
             \\    }
@@ -113,16 +129,16 @@ pub fn main() !void {
             \\
         ) catch {};
 
-        for (types) |typ| w.print(
+        for (event_types) |event_type| w.print(
             \\    {s}: {s},
             \\
-        , .{ lowercase(typ), typ }) catch {};
+        , .{ lowercase(event_type.name), event_type.name }) catch {};
 
-        inline for (types) |typ| {
+        inline for (event_types) |event_type| {
             w.print(
                 \\    pub const {s} = enum(c_ushort) {{
                 \\
-            , .{typ}) catch {};
+            , .{event_type.name}) catch {};
             defer w.print(
                 \\        pub inline fn new(integer: c_ushort) @This() {{
                 \\            return @enumFromInt(integer);
@@ -141,7 +157,7 @@ pub fn main() !void {
                 \\        }}
                 \\    }};
                 \\
-            , .{lowercase(typ)}) catch {};
+            , .{lowercase(event_type.name)}) catch {};
 
             const Alias = struct {
                 name: []const u8,
@@ -151,29 +167,36 @@ pub fn main() !void {
             defer aliases.deinit(allocator);
             var is_empty = true;
 
-            consts: inline for (consts) |cnst| {
-                comptime if (!filter(cnst, typ ++ "_"))
-                    if (!(std.mem.eql(u8, typ, "KEY") and filter(cnst, "BTN_")))
-                        continue;
-                comptime for (types) |t|
-                    if (!std.mem.eql(u8, typ, t) and filter(cnst, t ++ "_"))
-                        // - typ != t
-                        // - cnst starts with both typ and t
-                        if (typ.len < t.len)
-                            // t includes typ
-                            continue :consts;
+            var event_codes: std.ArrayList(Constant) = .empty;
+            defer event_codes.deinit(allocator);
+            if (std.mem.eql(u8, event_type.name, "KEY")) {
+                try event_codes.appendSlice(allocator, Constant.collect("KEY_"));
+                try event_codes.appendSlice(allocator, Constant.collect("BTN_"));
+            } else {
+                try event_codes.appendSlice(allocator, Constant.collect(event_type.name ++ "_"));
+            }
+            for (event_codes.items) |event_code| {
+                // Whether `event_code` should match a prefix that is longer than `event_type.name ++ "_"`.
+                const is_mismatched = for (event_types) |t| {
+                    if (t.name.len <= event_type.name.len) continue;
+                    if (!std.mem.startsWith(u8, t.name, event_type.name)) continue;
+                    if (std.mem.startsWith(u8, event_code.full_name, t.name)) break true;
+                } else false;
+                if (is_mismatched) continue;
 
                 is_empty = false;
 
-                const code: c_ushort = @field(c, cnst);
-                const code_name = if (c.libevdev_event_code_get_name(@field(c, "EV_" ++ typ), code)) |p| std.mem.span(p) else cnst;
-                if (std.mem.eql(u8, cnst, code_name))
+                const real_code_name = if (c.libevdev_event_code_get_name(
+                    event_type.value,
+                    @intCast(event_code.value),
+                )) |p| std.mem.span(p) else event_code.full_name;
+                if (std.mem.eql(u8, event_code.full_name, real_code_name))
                     w.print(
                         \\        {s} = {},
                         \\
-                    , .{ cnst, code }) catch {}
+                    , .{ event_code.full_name, event_code.value }) catch {}
                 else
-                    try aliases.append(allocator, .{ .name = cnst, .target = code_name });
+                    try aliases.append(allocator, .{ .name = event_code.full_name, .target = real_code_name });
             }
 
             for (aliases.items) |alias| w.print(

--- a/build/gen_event.zig
+++ b/build/gen_event.zig
@@ -51,7 +51,9 @@ const Constant = struct {
 
 pub fn main() !void {
     @setEvalBranchQuota(30000);
-    const allocator = std.heap.page_allocator;
+    var debug_allocator: std.heap.DebugAllocator(.{}) = .init;
+    defer std.debug.assert(debug_allocator.deinit() == .ok);
+    const allocator = debug_allocator.allocator();
 
     var s: std.ArrayList(u8) = .empty;
     defer s.deinit(allocator);

--- a/build/gen_event.zig
+++ b/build/gen_event.zig
@@ -33,6 +33,8 @@ const Constant = struct {
         return comptime b: {
             var slice: []const Constant = &.{};
             for (@typeInfo(c).@"struct".decls) |decl| {
+                // We need to filter declarations before accessing it with `@field()`
+                // to avoid `@compileError()`.
                 if (!std.mem.startsWith(u8, decl.name, prefix)) continue;
                 if (isMetaConstant(decl.name)) continue;
                 const self: Constant = .{
@@ -177,6 +179,7 @@ pub fn main() !void {
             }
             for (event_codes.items) |event_code| {
                 // Whether `event_code` should match a prefix that is longer than `event_type.name ++ "_"`.
+                // e.g.: `FF_STATUS_PLAYING` code should be matched with `EV_FF_STATUS` type instead of `EV_FF` type.
                 const is_mismatched = for (event_types) |t| {
                     if (t.name.len <= event_type.name.len) continue;
                     if (!std.mem.startsWith(u8, t.name, event_type.name)) continue;
@@ -220,15 +223,4 @@ pub fn main() !void {
     } else {
         std.debug.print("{s}", .{s.items});
     }
-}
-
-fn filter(name: []const u8, prefix: []const u8) bool {
-    if (!std.mem.startsWith(u8, name, prefix))
-        return false;
-
-    for ([_][]const u8{ "_VERSION", "_CNT", "_MAX" }) |s|
-        if (std.mem.endsWith(u8, name, s))
-            return false;
-
-    return true;
 }

